Sensor api example

zivsha opened the pull request.

shiritbrook committed:
Update readme.md

shiritbrook committed:
Update readme.md

Commit:
Fix unit-tests for D400/D410 and D5U.
Known issues:
 - D400(Passive) does not support all presets (that include laser power settings)
 - Pipeline test fail when multiple devices connected ( due to cycling with hub.wait_for_devices())

Change-Id: I4d6ac64e09624eda1ae4af8d082c557a50d4f7f1

icarpis committed:
Added back syncer class to work with callbacks

Change-Id: I65cc6fec2b41d9fe0dbe2f199e23c95fba7323db

Commit:
Fix pipeline device selection.
Tracked On: DSO-7326
Fix unit-test profiles for D430
remove whitespaces and resolve some compiler warnings

Change-Id: I749fd09872055ba1f7abf99be66a95b22d808464
Signed-off-by: Evgeni Raikhel <evgeni.raikhel@intel.com>

zivsha committed:
Fixed missing sensor->streaming=false for device models of disconnected devices

Commit:
Requesting a pipeline to allocate a device by serail will not invoke internal cycling
Fix compiler warnings - a potential bug in auto-complete class

Change-Id: Ia3bba8cf1988f02f962533672d8f7b5ea3c3805e

ev-mp committed:
Merge pull request #709 from zivsha/viewer_bug_fix

Bug fix when disconnecting camera while streaming

icarpis committed:
Fix rs-data-collect tool

Change-Id: I1769258abfdfecff70065e129e8e2db05a5fb206

ev-mp committed:
Merge pull request #705 from icarpis/syncer_with_callbacks

Added back syncer_processing_block - will be utilized in ROS wrapper

zivsha committed:
Started work on sensor api sample

dorodnic committed:
Merge pull request #710 from IntelRealSense/shiritbrook-patch-1

Shiritbrook patch 1

dorodnic committed:
Merge pull request #711 from icarpis/rs-data-collect_fix

Fix rs-data-collect tool

Commit:
Fine-tune wait_for_device implementation and adjust documentation

Change-Id: Iabc7ccd82286d133760b9ae4f79737a1fc16d3ed

Commit:
Modify timeout default values.
The default timeout duration must not be defined in terms of milliseconds::max
due to implementation bugs that overflow and produce negative durations.
The bug applies both to MSVC and GCC.
The alternative implementation will use chrono::hours::max which is 23 bits only.
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=58931
https://stackoverflow.com/questions/27726818/stdcondition-variablewait-for-exits-immediately-when-given-stdchronodura

Change-Id: I458dd588495c7f6d13134b265e4addd8df708447
Signed-off-by: Evgeni Raikhel <evgeni.raikhel@intel.com>

zivsha committed:
Additional documentation for open and start

Commit:
Fix issues profiled by CI

Change-Id: I8a026492bfca87808832e0f4303c6d0085460d50

aangerma committed:
Update pipeline - will raise only full sets of frames.

Commit:
Switching SR300 to use metadata for timestamp and frame number when available

Change-Id: I333f4c07ef6023a04d8fdfaf4fc56937bb98e806

dorodnic committed:
Fixing the issue where Advanced Mode controls were not showing in the Depth Quality tool

Commit:
Change advanced mode files to Unix/LF format. Clean up compiler warnings

Change-Id: I0809ea79e99924905acff07a3948ff1b31c15acf

zivsha committed:
Fixed pipline resolution to select the requested device when  streams were also requested

Commit:
SR300 to use metadata for timestamp/frame_id when available.
Fix recorder causing crash on application closure when trying to write to an already existing file

Change-Id: I7df054e25e3544ba701d39ebb61bdfb25d2a388a
Signed-off-by: Evgeni Raikhel <evgeni.raikhel@intel.com>

Commit:
API documentation fix

Change-Id: Ie12fe7c2f528ba601d06144bc982447ea6cedef7

zivsha committed:
Added options to sample, completed sensor streaming flow

icarpis committed:
Fixed Advanced Mode issue

Change-Id: I735d9c87e317098dda1f4297e1617d2e1c053658

zivsha committed:
Updated readme and some comments refactoring

ev-mp committed:
Merge pull request #712 from ev-mp/device_selection_by_pipeline

Rectify device selection for the pipeline

oordan commented on examples/sensor-control/readme.md:
```diff
@@ -0,0 +1,18 @@
+In addition to the `r2::pipeline` API, which simplifies streaming flow, the SDK provides the `rs2::context`, `rs2::device`, and `rs2::sensor` APIs that provide an abstraction for RealSense devices.
+
+In this sample we show how to use this API for querying all connected devices, their sensors, and their streams.
+Using a `rs2::sensor` one can control a camera's options, stream from it, and get the frames directly without having them synchronize (as the `rs2::pipeline` does).
```
without having them processed or synchronized.

oordan commented on examples/sensor-control/readme.md:
```diff
@@ -0,0 +1,18 @@
+## Overview
+
+The sensor control samples aims to provide the entry point and a tutorial for using the more advanced API that the RealSense SDK provides.
+In addition to the `r2::pipeline` API, which simplifies streaming flow, the SDK provides the `rs2::context`, `rs2::device`, and `rs2::sensor` APIs that provide an abstraction for RealSense devices.
```
I would link here to the API overview

zivsha replied:
which API overview?

oordan replied:
https://github.com/IntelRealSense/librealsense/blob/master/doc/api_arch.md
is there documentation of the sensor API design anywhere? if not I would add your comments to the api arch.

oordan commented on examples/sensor-control/readme.md:
```diff
@@ -0,0 +1,18 @@
+## Expected Output
+
+Running this example with a connected Intel RealSense device, will output the device name and some of its details.
+For each one of the device's sensors, the output will contain a list of supported stream profiles, and if all goes well a number of frames should arrive and print their number and stream type to theo utput.
```
the output

oordan commented on examples/sensor-control/rs-sensor-control.cpp:
```diff
@@ -0,0 +1,243 @@
+
+#include <librealsense2/rs.hpp>
+#include "../example.hpp"
+#include <imgui.h>
```
imgui and example are required dependencies here?

oordan commented on examples/sensor-control/rs-sensor-control.cpp:
```diff
@@ -0,0 +1,243 @@
+    //  connected devices
+    rs2::context ctx;
+
+    // Using the context we can get all connected devices into a device list
```
replace into with in

oordan commented on examples/sensor-control/rs-sensor-control.cpp:
```diff
@@ -0,0 +1,243 @@
+            {
+                // A Stream is an abstraction for a sequence of data items of a
+                //  single data type, which are ordered according to their time
+                //  of creation or arrival.
```
arrival?

zivsha replied:
I think this is originally your statement. do you want me to remove this ?

oordan replied:
I don't remember documenting this, but it confuses me now :)
do you think the order of frames can change between creation and arrival? it should be a bug if so, so I assume that the order is by creation. even if the TS is not...

oordan commented on examples/sensor-control/rs-sensor-control.cpp:
```diff
@@ -0,0 +1,243 @@
+                //  stream type we can use the following methods:
+
+                // 1) Each stream type can have multiple occurances.
+                //    All streams derived from a single device have distinc indices:
```
distinct. is it? don't color and depth have index 0?
derived? I would use "provided by"

oordan commented on examples/sensor-control/rs-sensor-control.cpp:
```diff
@@ -0,0 +1,243 @@
+                //     rather a human readable description of the stream
+                std::string stream_name = stream_profile.stream_name();
+
+                // 3) Each stream in the system, which derives from the same
```
derives from the same context? I thought that context always includes all devices. and you can instantiate multiple contexts. who gives the unique id? isn't it the singleton? 
derived is not a good word here. I think it's enough to say "each stream in the system / platform"

zivsha replied:
The context provides the unique ids, and it is not a singleton so if you have 2 contexts and you query the same device from them (I think this is supported) then you could have the same stream with 2 different ids.

oordan replied:
is this a feature or a bug?

zivsha replied:
I believe it was by design

oordan commented on examples/sensor-control/rs-sensor-control.cpp:
```diff
@@ -0,0 +1,243 @@
+                // 3) Each stream in the system, which derives from the same
+                //     rs2::context, has a unique identifier
+                //    This identifier is unique across all streams, and not
+                //     only between stream with the same stream type
```
this identifies is unique across all streams, regardless of the stream type.

oordan commented on examples/sensor-control/rs-sensor-control.cpp:
```diff
@@ -0,0 +1,243 @@
+
+                // As noted, a stream is an abstraction.
+                // In order to get additional data for the specific type of
+                //  stream, a mechanism of "Is" and "As" is provided:
```
type of *a* stream

zivsha replied:
Are you sure? sound right without the `a`...

oordan replied:
you are right, I meant to offer "for a specific type of stream", which also not required :). ignore.

oordan commented on examples/sensor-control/rs-sensor-control.cpp:
```diff
@@ -0,0 +1,243 @@
+            }
+
+            // The sensor controls turning the streaming on and off
+            // To being streaming, two calls must be made with the following order:
```
to begin, but I would write "to start"

oordan commented on examples/sensor-control/rs-sensor-control.cpp:
```diff
@@ -0,0 +1,243 @@
+            {
+                auto first_profile = *sensor.get_stream_profiles().begin();
+                // Open can be called with a single profile, or with a collection of profiles
+                // Calling open() tries to get exclusive access to the actual sensor (not only the software object of sensor)
```
to the physical sensor device. I would remove the (not only...).
explain here that the device may actually start running, produce data, and consume power.

ev-mp committed:
Merge pull request #714 from dorodnic/depth_quality_tool

Fixing the issue where Advanced Mode controls were not showing in the…

oordan commented on examples/sensor-control/rs-sensor-control.cpp:
```diff
@@ -0,0 +1,243 @@
+                //  * Since a sensor can stream multiple streams, and start
+                //     takes a single handler, multiple types of frames can arrive to the handler.
+                //  * Different streams arrive from different threads.
+                //    This behavior requires that the frame handler you provide to the start method, must be reentrant
```
This behavior requires the provided frame handler to the start method to be re-entrant

zivsha committed:
fixed warning and error for gcc compilation

oordan commented on examples/sensor-control/rs-sensor-control.cpp:
```diff
@@ -0,0 +1,243 @@
+                // NOTE:
+                //  * Since a sensor can stream multiple streams, and start
+                //     takes a single handler, multiple types of frames can arrive to the handler.
+                //  * Different streams arrive from different threads.
```
different streams frames arrive on different threads

oordan commented on examples/sensor-control/rs-sensor-control.cpp:
```diff
@@ -0,0 +1,243 @@
+                }
+
+                // After calling start, frames will begin to arrive asynchronously to the callback handler.
+                // We now block the main thread from continuing, to allows frames to arrive during this time
```
to allow

zivsha committed:
Typos and changes according to CR

Commit:
Update CI records to include D405 and SR300 in addition to D415/D435

Change-Id: I8faed1ba7bd1d9118d4db1af02b2e225ff5962d2
Signed-off-by: Evgeni Raikhel <evgeni.raikhel@intel.com>

zivsha committed:
Updated test

zivsha committed:
removed copy-pasted imgui leftovers

zivsha committed:
Updated main examples readme

zivsha committed:
A bit more documentation

aangerma committed:
Update pipeline - will raise only full sets of frames.

# Conflicts:
#	src/pipeline.cpp

aangerma committed:
Merge branch 'development' of https://github.com/IntelRealSense/librealsense into development

# Conflicts:
#	src/pipeline.cpp

aangerma committed:
Merge branch 'public_development' into development

aangerma committed:
added unsafe_stop() when restart device if device reconnected during streaming

aangerma committed:
Changes following code review

aangerma committed:
Update travis and appveyor to take new record files

ev-mp committed:
Merge pull request #717 from icarpis/advanced_mode_issue

Fixed issues in Advanced Mode and rs_data_collect tool
Add Depth White-Balance control to list of Advanced Mode attributes

aangerma committed:
pipeline::wait_for_frames - will not return empty frame holder
in case of camera reset.

aangerma committed:
pipeline - fix the comment of exception
"Device disconnected. Failed to recconect..."

zivsha committed:
Updated align sample to dynamically update align object in case of:
1 - device does not have color stream
2 - device disconnected and another one connected

aangerma committed:
Merge branch 'development' of https://github.com/IntelRealSense/librealsense into development

zivsha committed:
Added resolve() and then start() flow

icarpis committed:
Fixed issue in device_hub

Change-Id: I28059e1982d85aff77db427fd57445018153403a

zivsha committed:
changes to docs and names

ev-mp committed:
Merge pull request #716 from zivsha/pipeline_enable_stream_with_device

Pipeline config bug fix for user requesting a pipeline with a specific device (by serial number)  and enabling stream
Add unit-test for the above Use-Case.

aangerma committed:
Updated the version.

aangerma committed:
Added assert when non-composite frame arrived to pipeline.

ev-mp committed:
Merge pull request #723 from icarpis/development

Fixed issues in device_hub

aangerma committed:
Updated CI scripts

aangerma committed:
Remove remark

aangerma committed:
fix order of destruction in pipeline

zivsha committed:
Wrapped aggregator with queue to prevent dependency on pipeline (capturing 'this' in processor's lambda)

zivsha committed:
updated CI scripts to new recordings

zivsha committed:
split declaration and definition to make code cleaner.
Added return value validation for frames in pipeline processor

ev-mp committed:
Merge pull request #727 from zivsha/pipeline_processing_block

Pipeline behavior change - returned framesets will contain frames from all active streams

haoyunfeix committed:
Add test cases for nodejs classes depthSensor, frame and device

zivsha committed:
Updated readme after rebase

zivsha committed:
added expected output gif

icarpis committed:
Provide extrinsics calibration data in rs-enumerate-devices

Change-Id: I28adb0688772b825ea0396daf71c252e35428e0a

Commit:
Avoid set_xu(11) AE property failed when applied multiple times from different contexts.
Tracked on: DSO-7101

Change-Id: I60d042f67b3018e958fa61fbc8bd60f2b62e5553
Signed-off-by: Evgeni Raikhel <evgeni.raikhel@intel.com>

zivsha committed:
Restored missing 'break' statement inside unsafe_start retry mechanism

icarpis committed:
Fixed typo

Change-Id: Ia4d3ebe3b3450ac35feb5f1647307da005b7a284

zivsha committed:
Fixed enable_stream overload missing format

ev-mp committed:
Merge pull request #733 from icarpis/extrinsics_calibration_data

Provide extrinsic calibration data in rs-enumerate-devices
Fix presentation for profiles without intrinsic

ev-mp committed:
Merge pull request #734 from zivsha/pipeline_start_bug

Pipeline Bug fix

zivsha committed:
restored imgui and glfw

Commit:
Fix playback to preemptively select the next call

Change-Id: I0cfc29391c8d7bb35349196e863c616912a8863a
Signed-off-by: Evgeni Raikhel <evgeni.raikhel@intel.com>

Commit:
Modify Travis and Appvoyer settings with the record files for 2.8.1

Change-Id: I5edd9de1fab9e39c11022f1af7cd20c57bdd0f60

Commit:
Merge pull request #736 from haoyunfeix/add-nodejs-test

[Node.js] Add test cases for nodejs classes depthSensor, frame and device

Commit:
Make Appvoyer to complete all tests and only then report errors if any occurred
Use MSVC parallel build
Enforce Compilation with Metadata  (WinSDK 10.586)
Making Appvoyer to fail tests intentionally to test the new functionality

Change-Id: I4d7543b418b8e8c7e98e8f262555951488e9c49d
Signed-off-by: Evgeni Raikhel <evgeni.raikhel@intel.com>

Commit:
Switch AppVoyer fro test to operational mode

Change-Id: Id7dc13919fca82b7c1de146a7b7dd228a0060819

ev-mp committed:
Merge pull request #724 from zivsha/align_sample_update

Align example update

ev-mp committed:
Merge pull request #719 from ev-mp/ci_tests

Add D405 and SR300 recordings to CI's unit-tests suite
Split unit-tests records into groups for more effective management
Modify CI configuration to retrieve and run the tests from the different groups
Fix identified issues with recorder class utility

zivsha committed:
Added options control to gui
